File: backend/app/models/environment.rb
class Environment < ApplicationRecord
  belongs_to :organization
  has_many :environment_variables, dependent: :destroy
  has_many :environment_accesses, dependent: :destroy
  has_many :users, through: :environment_accesses
  
  validates :name, presence: true, uniqueness: { scope: :organization_id }
  validates :environment_type, presence: true, inclusion: { in: %w[development staging production test] }
  
  scope :for_organization, ->(org_id) { where(organization_id: org_id) }
  scope :by_type, ->(type) { where(environment_type: type) }
  
  def add_variable(key, value, encrypted: false)
    environment_variables.create!(
      key: key,
      value: value,
      encrypted: encrypted
    )
  end
  
  def get_variable(key)
    var = environment_variables.find_by(key: key)
    return nil unless var
    var.encrypted? ? decrypt_value(var.value) : var.value
  end
  
  def grant_access(user, role: 'viewer')
    environment_accesses.find_or_create_by(user: user) do |access|
      access.role = role
    end
  end
  
  def revoke_access(user)
    environment_accesses.where(user: user).destroy_all
  end
  
  def has_access?(user)
    environment_accesses.exists?(user: user)
  end
  
  def can_edit?(user)
    access = environment_accesses.find_by(user: user)
    access&.role == 'editor' || access&.role == 'admin'
  end
  
  private
  
  def decrypt_value(encrypted_value)
    EncryptionService.decrypt(encrypted_value)
  end
end

File: frontend/src/types/environment.ts
export interface Environment {
  id: number;
  name: string;
  environment_type: 'development' | 'staging' | 'production' | 'test';
  description?: string;
  organization_id: number;
  created_at: string;
  updated_at: string;
  environment_variables?: EnvironmentVariable[];
  environment_accesses?: EnvironmentAccess[];
}

export interface EnvironmentVariable {
  id: number;
  key: string;
  value: string;
  encrypted: boolean;
  description?: string;
  environment_id: number;
}

export interface EnvironmentAccess {
  id: number;
  role: 'viewer' | 'editor' | 'admin';
  user_id: number;
  environment_id: number;
  user?: {
    id: number;
    email: string;
    name: string;
  };
}

